From 8f8d2537a2401774be6ed8e1d9510844eb80e798 Mon Sep 17 00:00:00 2001
From: Charlie Turner <cturner@igalia.com>
Date: Thu, 1 Feb 2018 17:11:31 +0000
Subject: [PATCH] Remove closed sessions from session id map

This isn't a final fix, the session id handling is completely bonkers... But
this does fix the issue of gstreamer selecting incorrect sessions
non-determinisically IFF session.close() is called by the application.
Multi-key will be refactoring much of this session handling stuff.

* platform/graphics/gstreamer/eme/CDMOpenCDM.cpp:
(WebCore::CDMInstanceOpenCDM::loadSession):
(WebCore::CDMInstanceOpenCDM::closeSession):
---
 Source/WebCore/platform/graphics/gstreamer/eme/CDMOpenCDM.cpp | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/Source/WebCore/platform/graphics/gstreamer/eme/CDMOpenCDM.cpp b/Source/WebCore/platform/graphics/gstreamer/eme/CDMOpenCDM.cpp
index d569fb006da..a2cf71ac758 100644
--- a/Source/WebCore/platform/graphics/gstreamer/eme/CDMOpenCDM.cpp
+++ b/Source/WebCore/platform/graphics/gstreamer/eme/CDMOpenCDM.cpp
@@ -407,14 +407,18 @@ void CDMInstanceOpenCDM::loadSession(LicenseType, const String& sessionId, const
     }
 }
 
-void CDMInstanceOpenCDM::closeSession(const String&, CloseSessionCallback callback)
+void CDMInstanceOpenCDM::closeSession(const String& sessionId, CloseSessionCallback callback)
 {
+    if (!sessionIdMap.remove(sessionId)) {
+        GST_WARNING("%s is an unknown session", sessionId.utf8().data());
+    }
     m_openCdmBackend->Close();
     callback();
 }
 
 void CDMInstanceOpenCDM::removeSessionData(const String& sessionId, LicenseType, RemoveSessionDataCallback callback)
 {
+    // FIXME: Should the session id be removed from the map here as well as in closeSession?
     std::string responseMessage;
     KeyStatusVector keys;
     int ret = m_openCdmBackend->Remove(responseMessage);
-- 
2.15.1

